src/InlineAutocomplete/utils.ts

```diff
@@ -0,0 +1,156 @@
+import getCaretCoordinates from '@koddsson/textarea-caret'
+import {Children, EventHandler, SyntheticEvent} from 'react'
+
+import {Coordinates, ShowSuggestionsEvent, Suggestion, TextInputCompatibleChild, Trigger} from './types'
+
+const singleWordTriggerTerminators = new Set([' ', '\n'])
+const multiWordTriggerTerminators = new Set(['.', '\n'])
+
+const isWhitespace = (char: string) => /\s/.test(char)
+
+/**
+ * Calculate whether or not suggestions should be shown based on the given state of the
+ * input. If they should be shown, returns the show event.
+ */
+export const calculateSuggestionsQuery = (
+  triggers: Array<Trigger>,
+  text: string,
+  caretLocation: number
+): ShowSuggestionsEvent | null => {
+  // Build backwards from the caret location until the most recent trigger character or terminator
+  for (
+    let i = caretLocation - 1, query = '', potentialTriggers = triggers;
+    i >= 0 && potentialTriggers.length > 0;
+    i--
+  ) {
+    const character = text[i]
+
+    if (singleWordTriggerTerminators.has(character)) potentialTriggers = potentialTriggers.filter(t => t.multiWord)
+    if (multiWordTriggerTerminators.has(character)) potentialTriggers = potentialTriggers.filter(t => !t.multiWord)
+
+    for (const trigger of potentialTriggers.filter(t => character === t.triggerChar)) {
+      // Trigger chars must always be preceded by whitespace or be the first character in the input,
+      // and even a multi-word query cannot start with whitespace
+      if ((i === 0 || isWhitespace(text[i - 1])) && !isWhitespace(query[0])) return {trigger, query}
+
+      potentialTriggers = potentialTriggers.filter(t => t !== trigger)
+    }
+
+    query = character + query
+  }
+
+  return null
+}
+
+/**
+ * Obtain the coordinates (px) of the bottom left of a character in an input, relative to the
+ * top-left corner of the input itself.
+ * @param input The target input element.
+ * @param index The index of the character to calculate for.
+ * @param adjustForScroll Control whether the returned value is adjusted based on scroll position.
+ */
+export const getCharacterCoordinates = (
+  input: HTMLTextAreaElement | HTMLInputElement | null,
+  index: number,
+  adjustForScroll = true
+): Coordinates => {
+  if (!input) return {top: 0, left: 0}
+
+  // word-wrap:break-word breaks the getCaretCoordinates calculations (a bug), and word-wrap has
```

Comment:
Thanks for working around these `getCaretCoordinates` bugs. It would be nice to fix them in the library, but that's definitely not a blocker for this PR.